### ISSUE TITLE
ZMQ.Socket obtained from ZContext behaves differently than ZMQ.Socket obtained from ZMQ.Context

### ISSUE THREAD

**Issue by Koshroy** [zeromq/jeromq#140]
A ZMQ.Socket obtained from a ZContext object does not need to be explicitly closed as the ZContext manager manages the liveness of the ZMQ.Socket. However, a ZSocket obtained from a ZMQ.Context does need to be explicitly closed. The ZContext will refuse to destroy cleanly if any of the ZMQ.Socket objects are user closed. This seems counter-intuitive, as both context objects return ZMQ.Socket objects. Either a ZContext should return a high-level wrapper around a ZMQ.Socket (ZSocket maybe?) that lets the Java user understand the differences in semantics, or ZContext should detect when a ZMQ.Socket has been explicitly closed and not block indefinitely trying to close it.

**Comment by daveyarwood** [comment]
I'm late to the party, but I took a look at @trevorbernard's ZSocket commit above and it LGTM, but I'm confused about how it relates to the [ZSocket](https://github.com/zeromq/jeromq/blob/master/src/main/java/org/zeromq/ZSocket.java) on master added in a much later commit (January 2016). Seems like the intent is the same. Does that mean this issue can be closed?

On a more general note, it seems like JeroMQ could use better documentation, or perhaps more examples, illustrating the different classes available for using contexts/sockets and maybe even pointing the user toward the latest recommended way. The ZMQ guide examples seem to all involve either using `ZMQ.context(1)` or `new ZContext()` to create the context, and then manually destroying it when done. As a user I find this confusing, and it makes me wonder which way is the "right" way and which way might be deprecated. Do we want to recommend using ZSockets and not dealing with the context directly?

**Comment by daveyarwood** [comment]
I also noticed that ZPoller apparently cannot be used with ZSockets, as [its `register` method takes a Socket](https://github.com/zeromq/jeromq/blob/master/src/main/java/org/zeromq/ZPoller.java#L328-L336) and not a ZSocket. Overall I could use some guidance on how to use the new "ZThings." I'd be willing to help improve documentation based on what I learn.

**Comment by daveyarwood** [comment]
Looking further into ZPoller and the other ZThings, it seems to me that in general, these things have not been fully implemented or perhaps even fully thought through / spec'd. Case in point: there are no "real" tests for ZPoller. The ones in [TestZPoller](https://github.com/zeromq/jeromq/blob/master/src/test/java/org/zeromq/TestZPoller.java) seem to be testing the behavior when you try to use ZPoller with null contexts, sockets, channels, etc. Contrast this with the TestZMQ tests, which use ZMQ.Pollers and test things like [whether POLLOUT works](https://github.com/zeromq/jeromq/blob/master/src/test/java/org/zeromq/TestZMQ.java#L46) when used with an actual (non-null) context and sockets.

I think what we have on master is a step in the right direction, but more work is needed before we can recommend using any new API where the ZMQ context is hidden. At the moment, I think creating a ZContext and using that to manage your sockets/pollers is the way to go.

One issue we need to keep in mind is that even if the context is hidden from the user, there needs to be a way to shut down all of the things at will. The context is convenient for this because for example, when you terminate a ZContext, it ensures that all sockets are closed. We also need a way to make sure that polling resources (e.g. Selectors) are shut down when done.

One approach we could take is to have the hidden context continue to manage sockets, polling selectors, etc., and expose a hypothetical method like `ZMQ.shutdown`, which shuts down the managed context along with all registered resources.

We can probably do this in a backwards-compatible way, so that you can either explicitly create your own context or use the managed one.

In the interest of keeping this moving, I'm going to create the following issues:

- New issue: complete the work of this "hidden context" idea, so that we can move toward a new contextless API. I have some thoughts about this, which I can capture in the issue.

- New issue: we need better ZPoller tests

- New issue: ZPoller can't register ZSockets

**Comment by daveyarwood** [comment]
> The ZContext will refuse to destroy cleanly if any of the ZMQ.Socket objects are user closed.

Of note, I think #388 fixed this behavior, so this issue can be closed. The problem was that when you destroy a ZContext, it goes through all of its sockets and runs `setLinger(0)` followed by `close`, however prior to #388 being merged, you could not call `setLinger` on a closed socket without an error being thrown.
